File: MinFallPathSum.java
// Time Complexity : O(n^2) (since the rows and columns are equal)
// Space Complexity : O(1) (since we are modifying the orginal array without using DP array)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/* Greedy approach fails so we use bottom up dynamic programming.
This is similar to house coloring problem.
Start from last but on row and add the minimum of the below two possible elements until we reach the first row.
First row has the sum of the path  and we take the minimum of first row. 
 * 
*/
class Solution {
    public int minFallingPathSum(int[][] matrix) {
        if(matrix == null || matrix.length == 0) return 0;
        int m = matrix.length;
        //bottom up approach
        //so start adding from last but one row
        for(int i = m-2;i>=0;i--){
            for(int j=0;j<m;j++){
                if(j==0){//if the element is at the first column then it wpuld not have left diagonal
                    matrix[i][j] = matrix[i][j] + Math.min(matrix[i+1][j],matrix[i+1][j+1]);
                }
                else if(j==m-1){//if the element is at the last column then it wpuld not have right diagonal
                    matrix[i][j] = matrix[i][j]+ Math.min(matrix[i+1][j],matrix[i+1][j-1]);
                }
                else{//if its the column between first and last
               matrix[i][j] = matrix[i][j]+ Math.min(matrix[i+1][j-1],Math.min(matrix[i+1][j],matrix[i+1][j+1]));
                }
            }
        }
 int min = Integer.MAX_VALUE;
 //iterate through the first row and add the minimum.
        for(int j=0;j<m;j++){
            min = Math.min(min,matrix[0][j]);
        }
        return min;
    }
}

File: min_fall_path_sum.cpp
// Time Complexity: O(N*M)Where N is no. of rows in matrix and M is the no. of colums in matrix
// Space Complexity: O(N*M) Where N is no. of rows in matrix and M is the no. of colums in matrix
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        if(matrix.size()==0)return 0;
        int rows = matrix.size();
        int cols = matrix[0].size();
        int t[rows][cols];
        for(int i = 0; i < cols;i++)t[0][i]=matrix[0][i];
        
        for(int i = 1; i < rows;i++){
            for(int j = 0; j < cols;j++){
                if(j==0)t[i][j] = matrix[i][j] + min(t[i-1][0],t[i-1][1]);
                else if(j==cols-1)t[i][j] = matrix[i][j] + min(t[i-1][j],t[i-1][j-1]);
                else t[i][j] = matrix[i][j] + min_three(t[i-1][j],t[i-1][j-1],t[i-1][j+1]);                
            }
        }
        int min = INT_MAX;
        for(int i = 0; i < cols;i++){
            if(t[rows-1][i]<min)min = t[rows-1][i];
        }
        return min;
        
    }
    
    int min_three(int a,int b,int c){
        return min(min(a,b),c);
    }
};